clang/docs/checkedc/Bounds-Declaration-Checking.md

```diff
@@ -0,0 +1,208 @@
+  of `bounds(p, p + 5)`, then the rvalue expression from reading the value of
+  `*(p + 2)` has rvalue bounds of `bounds(unknown)`. `*(p + 2)` has type `int`,
+  and integers do not have bounds.
+- **LValue bounds**: The bounds of an lvalue expression `e`. These bounds
+determine whether it is valid to access memory using `e`, and should be the
+range (or a subrange) of an object in memory. LValue bounds are used to check
+that a memory access is within bounds.
+  - Example: If `p` is a non-array-typed variable, then the lvalue bounds
+  of `p` are `bounds(&p, &p + 1)`.
+  - Example: If `p` is a variable with declared bounds of `bounds(p, p + 3)`,
+  then `*(p + 1)` and `p[2]` have lvalue bounds of `bounds(p, p + 3)`. The
+  compiler will check that `p + 1` and `p + 2` are within the lvalue bounds of
+  `bounds(p, p + 3)`.
+- **Target bounds**: The target bounds of an lvalue expression `e`. Values
+assigned to `e` must satisfy these bounds. Values read through `e` will
+meet these bounds.
+  - Example: If `p` is a variable with declared bounds of `bounds(p, p + 1)`,
+  then the target bounds of `p` are `bounds(p, p + 1)`.
+  - Example: If `p` is a variable of type `array_ptr<int>` with declared
+  bounds of `bounds(p, p + 7)`, then the target bounds of `p[4]` are
+  `bounds(unknown)`.
+  - Example: If S is a struct with members
+  `{ array_ptr<int> f : count(len); int len; }`, then `s.f` has target bounds
+  of `bounds(s.f, s.f + s.len)`. In an assignment `s.f = e;`, the rvalue bounds
+  of `e` must imply `bounds(s.f, s.f + s.len)`.
+
+Examples of declared and inferred bounds:
+```
+// p has declared bounds of bounds(p, p + i).
+// q has declared bounds of bounds(q, q + j).
+void f(array_ptr<int> p : count(i), array_ptr<int> q : count(j)) {
+  // Updating after assignment: The inferred bounds of the left-hand side
+  // variable are the bounds of the right-hand side expression.
+  // p has inferred bounds of bounds(q, q + j).
+  p = q;
+
+  // Updating after an invertible modification to a variable used in declared
+  // bounds: The original value of i before this assignment is i - 1.
+  // This original value is substituted for i in the inferred bounds
+  // of p, so the inferred bounds of p are bounds(p, p + i - 1).
+  i = i + 1;
+
+  // Updating after some uninvertible modifications to a variable used in
+  // declared bounds: The compiler cannot determine an original value for i
+  // after these assignments, so the inferred bounds of p are bounds(unknown).
+  i = 0;
+  i = p[1] / 3;
+  i = *q;
+  i = 2 * i;
+}
+```
+
+## Bounds Validity
+After checking each statement in the clang CFG, the compiler attempts to prove
+or disprove that the inferred bounds of each variable imply the target bounds
+of the variable. (The target bounds of a variable are always the declared
+bounds of the variable). In addition, after an assignment to a non-variable
+lvalue, the compiler attempts to prove that the inferred bounds of the
```

Comment:
For a variable lvalue `x` (i.e. `x` is a `DeclRefExpr`), the check that the inferred bounds of `x` imply the target bounds of `x` happens after checking the current top-level CFG statement in the `TraverseCFG` method. For an lvalue `v` that isn't a variable (e.g. `v`  is `s.f`, `s->f`, `*p`, etc.), the check that the inferred bounds of `v` imply the target bounds of `v` happens at the end of checking an assignment to `v` in the `CheckBinaryOperator` method. After generalizing the bounds context checking work from variables to lvalue expressions, this should no longer be the case - all kinds of lvalue expressions should be bounds checked after checking the top-level CFG statement. I could take this sentence out since it will eventually be inaccurate, or I could add some examples of lvalues that are not variables for now and take the sentence out after the work to generalize from variables to lvalues is done.